File: flask_app/app/controllers/category_controller.py
from flask import request
from flask_jwt_extended import jwt_required
from app.services.category_service import CategoryService
from app.utils.response import success_response, error_response
from app.utils.decorators import admin_required

class CategoryController:

    @staticmethod
    @jwt_required()
    def get_all():
        page = request.args.get('page', default=1, type=int)
        per_page = request.args.get('per_page', default=10, type=int)
        filters = {
            "is_active": request.args.get("active", type=lambda v: v.lower() == 'true' if v else None),
        }

        filters = {k: v for k, v in filters.items() if v is not None}

        tiers, total = CategoryService.get_all(page, per_page)
        return success_response("Categories retrieved", {
            "items": [t.to_dict() for t in tiers],
            "total": total,
            "page": page,
            "per_page": per_page,
        })
    
    @staticmethod
    @jwt_required()
    def get_one(category_id):
        category = CategoryService.get_by_id(category_id)
        if not category:
            return error_response("Category not found", 404)
        return success_response("Category found", category.to_dict())

    @staticmethod
    @admin_required()
    def create():
        data = request.get_json()
        category = CategoryService.create(data)
        return success_response("Category created", category.to_dict())

    @staticmethod
    @admin_required()
    def update(category_id):
        data = request.get_json()
        updated_category = CategoryService.update(category_id, data)
        if not updated_category:
            return error_response("Category not found", 404)
        return success_response("Category updated", updated_category.to_dict())

    @staticmethod
    @admin_required()
    def delete(category_id):
        deleted = CategoryService.soft_delete(category_id)
        if not deleted:
            return error_response("Category not found", 404)
        return success_response("Category deleted successfully")

File: springboot_api/src/main/java/com/example/membership_api/controller/CategoryController.java
package com.example.membership_api.controller;

import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.example.membership_api.annotations.AdminOnly;
import com.example.membership_api.model.Category;
import com.example.membership_api.service.CategoryService;
import com.example.membership_api.utils.ResponseUtil;

@RestController
@RequestMapping("/categories")
public class CategoryController {

    @Autowired
    private CategoryService service;

    @GetMapping
    public ResponseEntity<?> getAll(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int perPage,
            @RequestParam(name = "active", required = false, defaultValue = "true") boolean onlyActive
    ) {
        Map<String, Object> data = service.getAll(page, perPage, onlyActive);
        return ResponseUtil.success("Categories retrieved", data);
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getOne(@PathVariable String id) {
        Category category = service.getById(id);
        return ResponseUtil.success("Category found", category);
    }

    @PostMapping
    @AdminOnly
    public ResponseEntity<?> create(@RequestBody Category category) {
        Category created = service.create(category);
        return ResponseUtil.success("Category created", created);
    }

    @PutMapping("/{id}")
    @AdminOnly
    public ResponseEntity<?> update(@PathVariable String id, @RequestBody Category updates) {
        Category updated = service.update(id, updates);
        return ResponseUtil.success("Category updated", updated);
    }

    @DeleteMapping("/{id}")
    @AdminOnly
    public ResponseEntity<?> delete(@PathVariable String id) {
        service.softDelete(id);
        return ResponseUtil.success("Category deleted successfully");
    }
}
